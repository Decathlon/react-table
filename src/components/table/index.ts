export { default as Table } from "./table";
export * from "./table";

export { default as Row } from "./row";
export * from "./row";

export { default as RowSpan } from "./row-span";
export * from "./row-span";

export { default as Cell } from "./cell";
export * from "./cell";

export { default as ElementaryTable } from "./elementary-table";
export * from "./elementary-table";
