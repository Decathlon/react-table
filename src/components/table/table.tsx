import * as React from "react";
import { isEqual } from "lodash";
import ElementaryTable, { IElementaryTable, IColumnOptions, ITree, IColumns, ITrees } from "./elementary-table";
import { IRowOptions } from "./row";
import ResponsiveContainer, { IResponsiveContainerOptionalProps } from "../responsive-container";
import Virtualizer, { IVirtualizerOptionalProps } from "../virtualizer";
import {
  getTreesLength,
  getAllIndexesMap,
  IIndexesMap,
  getFixedElementFixedSizeSum,
  relativeToAbsoluteIndexes,
  getIndexesIdsMapping,
  IIndexesIdsMapping,
  getColumnsLength,
  getCellPath,
  getCell,
  IRelativeIndexesMap,
  IElevateds,
  getDenseColumns
} from "../utils/table";
import SelectionHandler, { ISelection, ISelectionHandlerOptionalProps } from "../table-selection/selection-handler";
import { ROW_SPAN_WIDTH } from "../constants";
import { ICell, ICellCoordinates } from "./cell";
import shallowEqual from "../utils/shallowEqual";

interface IVirtualizerProps extends Partial<IVirtualizerOptionalProps> {
  /**  The width of the visible window. To specify if not responsive */
  width?: number;
  /**  The height of the visible window. To specify if not responsive */
  height?: number;
}

/**
 * We will be using something as close as possible to the original table,
 * with separated abstractions in this component
 * */
export interface ITableProps<IDataCoordinates = any> extends IElementaryTable<IDataCoordinates> {
  responsiveContainerProps: IResponsiveContainerOptionalProps;
  virtualizerProps: IVirtualizerProps;
  selectionProps?: ISelectionHandlerOptionalProps;
  isSelectable?: boolean;
  isVirtualized?: boolean;
  /** A list of branches to initialize opened rows and sub rows */
  initialOpenedTrees: ITrees;
}

export interface IState {
  rowsLength: number;
  indexesMapping: IIndexesMap;
  columnsIndexesIdsMapping: IIndexesIdsMapping;
  openedTrees: ITrees;
  fixedRowsIndexes: number[];
}

class Table<IDataCoordinates = any> extends React.Component<ITableProps<IDataCoordinates>, IState> {
  static defaultProps = {
    isSelectable: true,
    isVirtualized: false,
    virtualizerProps: {},
    responsiveContainerProps: {},
    initialOpenedTrees: {}
  };

  private globalRowProps?: IRowOptions;

  private globalColumnProps?: IColumnOptions;

  private fixedCellsHeight?: {
    sum: number;
    count: number;
  };

  private fixedCellsWidth?: {
    sum: number;
    count: number;
  };

  private virtualizer: React.RefObject<Virtualizer> = React.createRef<Virtualizer>();

  private columnsLength = 0;

  public constructor(props: ITableProps<IDataCoordinates>) {
    super(props);
    const {
      initialOpenedTrees,
      rows,
      columns,
      isVirtualized,
      virtualizerProps: { fixedRows, fixedColumns, hiddenRows, hiddenColumns }
    } = this.props;
    this.columnsLength = getColumnsLength(rows);
    const indexesMapping = getAllIndexesMap(initialOpenedTrees, rows);
    this.state = {
      indexesMapping,
      openedTrees: initialOpenedTrees,
      rowsLength: this.getRowslength(initialOpenedTrees),
      columnsIndexesIdsMapping: rows[0] ? getIndexesIdsMapping(rows[0].cells) : {},
      fixedRowsIndexes: this.getFixedRowsIndexes(initialOpenedTrees, indexesMapping.relative)
    };
    if (isVirtualized) {
      this.fixedCellsHeight = getFixedElementFixedSizeSum(rows, fixedRows, hiddenRows);
      this.fixedCellsWidth = getFixedElementFixedSizeSum(columns, fixedColumns, hiddenColumns);
    }
  }

  public shouldComponentUpdate(nextProps: ITableProps<IDataCoordinates>, nextState: IState) {
    const { virtualizerProps, initialOpenedTrees, ...otherProps } = this.props;
    const { virtualizerProps: nextVirtualizerProps, initialOpenedTrees: nextInitialOpenedTrees, ...nextOtherProps } = nextProps;
    const { initialScroll, ...otherVirtualizerProps } = virtualizerProps;
    const { initialScroll: nextInitialScroll, ...nextOtherVirtualizerProps } = nextVirtualizerProps;
    return (
      !shallowEqual(this.state, nextState) ||
      !shallowEqual(otherProps, nextOtherProps) ||
      !shallowEqual(otherVirtualizerProps, nextOtherVirtualizerProps)
    );
  }

  public componentDidUpdate(prevProps: ITableProps<IDataCoordinates>) {
    const { rows } = this.props;
    const { openedTrees } = this.state;
    if (prevProps.rows !== rows) {
      this.columnsLength = getColumnsLength(rows);
      const newIndexesMapping = getAllIndexesMap(openedTrees, rows);
      this.setState({
        indexesMapping: newIndexesMapping,
        rowsLength: this.getRowslength(openedTrees),
        columnsIndexesIdsMapping: rows[0] ? getIndexesIdsMapping(rows[0].cells) : {},
        fixedRowsIndexes: this.getFixedRowsIndexes(openedTrees, newIndexesMapping.relative)
      });
    }
  }

  private getFixedRowsIndexes = (openedTrees: ITrees, relativeIndexesMapping: IRelativeIndexesMap) => {
    const {
      rows,
      virtualizerProps: { fixedRows }
    } = this.props;
    const newfixedRowsIndexes = (fixedRows && relativeToAbsoluteIndexes(fixedRows, relativeIndexesMapping)) || [];
    return Object.keys(openedTrees).reduce<number[]>((result, rowIndex) => {
      const { fixSubRows } = rows[rowIndex];
      if (fixSubRows) {
        const { subItems } = relativeIndexesMapping[rowIndex];
        const rowSubItems = subItems || {};
        result.push(...Object.keys(rowSubItems).map(subItem => rowSubItems[subItem].index));
      }
      return result;
    }, newfixedRowsIndexes);
  };

  private getColumnsProps = (cellWidth: number) => {
    const { globalColumnProps } = this.props;
    const newGlobalColumnProps = { ...globalColumnProps, size: cellWidth };
    if (!isEqual(this.globalColumnProps, newGlobalColumnProps)) {
      this.globalColumnProps = newGlobalColumnProps;
    }
    return this.globalColumnProps;
  };

  private getRowsProps = (cellHeight: number) => {
    const { globalRowProps } = this.props;
    const newGlobalRowProps = { ...globalRowProps, size: cellHeight };
    if (!isEqual(this.globalRowProps, newGlobalRowProps)) {
      this.globalRowProps = newGlobalRowProps;
    }
    return this.globalRowProps;
  };

  private getRowslength = (openedTrees: ITrees): number => {
    const { rows } = this.props;
    let rowsLength = (rows && rows.length) || 0;
    rowsLength += openedTrees ? getTreesLength(openedTrees, rows) : 0;
    return rowsLength;
  };

  private updateRowsLength = (openedTrees: ITrees) => {
    const { rows } = this.props;
    const newIndexesMapping = getAllIndexesMap(openedTrees, rows);
    const newRowsLength = this.getRowslength(openedTrees);

    this.setState({
      indexesMapping: newIndexesMapping,
      openedTrees,
      rowsLength: newRowsLength,
      fixedRowsIndexes: this.getFixedRowsIndexes(openedTrees, newIndexesMapping.relative)
    });
  };

  private onRowOpen = (openedTree: ITree) => {
    const { openedTrees } = this.state;
    const currentOpenedTrees = openedTrees || {};
    const newOpenedTrees = { ...currentOpenedTrees };
    // update the sub-tree to close
    newOpenedTrees[openedTree.rowIndex] = openedTree;
    this.updateRowsLength(newOpenedTrees);
  };

  private onRowClose = (closedTree: ITree) => {
    const { openedTrees } = this.state;
    const currentOpenedTrees = openedTrees || {};
    const newOpenedTrees = { ...currentOpenedTrees };
    // remove the sub-tree to close
    delete newOpenedTrees[closedTree.rowIndex];
    if (Object.keys(newOpenedTrees).length !== Object.keys(openedTrees).length) {
      this.updateRowsLength(newOpenedTrees);
    }
  };

  public openTrees = (trees: ITrees) => {
    const { openedTrees } = this.state;
    const currentOpenedTrees = openedTrees || {};
    const newOpenedTrees = { ...currentOpenedTrees, ...trees };
    this.updateRowsLength(newOpenedTrees);
  };

  public closeTrees = (trees: ITrees) => {
    const { openedTrees } = this.state;
    const currentOpenedTrees = openedTrees || {};
    const newOpenedTrees = { ...currentOpenedTrees };
    Object.keys(trees).forEach(rowId => {
      delete newOpenedTrees[rowId];
    });
    this.updateRowsLength(newOpenedTrees);
  };

  public goToColumnIndex = (columnIndex: number) => {
    if (this.virtualizer.current) {
      const toColumnIndex = Math.max(Math.min(columnIndex, this.columnsLength - 1), 0);
      this.virtualizer.current.scrollToColumnIndex(toColumnIndex);
    }
  };

  public goToRowIndex = (rowIndex: number) => {
    if (this.virtualizer.current) {
      const { rowsLength } = this.state;
      const toRowIndex = Math.max(Math.min(rowIndex, rowsLength - 1), 0);
      this.virtualizer.current.scrollToRowIndex(toRowIndex);
    }
  };

  public goToColumnId = (columnId: string) => {
    if (this.virtualizer) {
      const columnIndex = this.getColumnIndex(columnId);
      if (columnIndex !== undefined) {
        this.goToColumnIndex(columnIndex);
      }
    }
  };

  public getColumnIndex = (columnId: string) => {
    const { columnsIndexesIdsMapping } = this.state;
    return columnsIndexesIdsMapping[columnId];
  };

  public getColumnId = (columnIndex: number) => {
    const { rows } = this.props;
    const header = rows[0] && rows[0].cells;
    const cell = header && header[columnIndex];
    return cell && cell.id;
  };

  public getCell = (cellCoordinates: ICellCoordinates): ICell<IDataCoordinates> => {
    const { rows } = this.props;
    const { indexesMapping, openedTrees } = this.state;
    const cellPath = getCellPath(cellCoordinates, indexesMapping.absolute, openedTrees);
    return getCell(rows, cellPath);
  };

  private renderTable = (
    visibleColumnIndexes?: number[],
    visibleRowIndexes?: number[],
    cellHeight?: number,
    cellWidth?: number,
    elevatedColumnIndexes?: IElevateds,
    elevatedRowIndexes?: IElevateds,
    fixedRowsIndexes?: number[],
    columns?: IColumns
  ) => {
    const {
      globalRowProps,
      globalColumnProps,
      isSelectable,
      selectionProps,
      responsiveContainerProps,
      virtualizerProps,
      isVirtualized,
      initialOpenedTrees,
      ...tableProps
    } = this.props;
    const { indexesMapping, openedTrees } = this.state;
    const renderElementaryTable = (selection: ISelection = { selectedCells: {} }): JSX.Element => (
      <ElementaryTable
        {...tableProps}
        {...selection}
        columns={columns || tableProps.columns}
        visibleColumnIndexes={visibleColumnIndexes || tableProps.visibleColumnIndexes}
        visibleRowIndexes={visibleRowIndexes || tableProps.visibleRowIndexes}
        fixedRowsIndexes={fixedRowsIndexes}
        globalRowProps={cellHeight ? this.getRowsProps(cellHeight) : globalRowProps}
        globalColumnProps={cellWidth ? this.getColumnsProps(cellWidth) : globalColumnProps}
        elevatedColumnIndexes={elevatedColumnIndexes || tableProps.elevatedColumnIndexes}
        elevatedRowIndexes={elevatedRowIndexes || tableProps.elevatedRowIndexes}
        onRowOpen={this.onRowOpen}
        onRowClose={this.onRowClose}
        indexesMapping={indexesMapping}
        openedTrees={openedTrees}
      />
    );
    return isSelectable ? (
      <SelectionHandler {...selectionProps}>{renderElementaryTable}</SelectionHandler>
    ) : (
      renderElementaryTable()
    );
  };

  private renderVirtualizedTable = (height: number, width: number) => {
    const {
      isSpan,
      virtualizerProps,
      virtualizerProps: { hiddenRows },
      columns
    } = this.props;
    const { rowsLength, indexesMapping, fixedRowsIndexes } = this.state;

    return (
      <Virtualizer
        ref={this.virtualizer}
        {...virtualizerProps}
        fixedRows={fixedRowsIndexes}
        columnsLength={this.columnsLength}
        hiddenRows={hiddenRows && relativeToAbsoluteIndexes(hiddenRows, indexesMapping.relative)}
        rowsLength={rowsLength}
        width={width}
        height={height}
        fixedCellsHeight={this.fixedCellsHeight}
        fixedCellsWidth={this.fixedCellsWidth}
        verticalPadding={isSpan ? ROW_SPAN_WIDTH : 0}
      >
        {({ visibleColumnIndexes, visibleRowIndexes, elevatedColumnIndexes, elevatedRowIndexes, cellHeight, cellWidth }) => {
          const tableWidth = visibleColumnIndexes.length * cellHeight;
          const denseColumns = getDenseColumns(tableWidth, width, this.columnsLength, columns);
          return this.renderTable(
            visibleColumnIndexes,
            visibleRowIndexes,
            cellHeight,
            cellWidth,
            elevatedColumnIndexes,
            elevatedRowIndexes,
            fixedRowsIndexes,
            denseColumns
          );
        }}
      </Virtualizer>
    );
  };

  private renderResponsiveTable = () => {
    const {
      responsiveContainerProps: { className }
    } = this.props;
    return (
      <ResponsiveContainer className={className}>
        {({ width, height }) => {
          return this.renderVirtualizedTable(height, width);
        }}
      </ResponsiveContainer>
    );
  };

  public render() {
    const {
      virtualizerProps: { height, width },
      isVirtualized
    } = this.props;
    if (isVirtualized) {
      if (height && width) {
        return this.renderVirtualizedTable(height, width);
      }
      return this.renderResponsiveTable();
    }
    return this.renderTable();
  }
}

export default Table;
