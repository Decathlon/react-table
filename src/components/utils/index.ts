export * from "./table";
export * from "./table-selection";
export * from "./common";

export { default as shallowEqual } from "./shallowEqual";
