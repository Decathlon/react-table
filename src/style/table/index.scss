@import "cell";
@import "row";
@import "row-span";
