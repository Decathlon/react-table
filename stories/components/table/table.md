# Table

Default table for your application.

Able to handle customized content, different row/column size, and child rows.
