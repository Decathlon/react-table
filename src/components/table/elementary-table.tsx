import * as React from "react";
import classnames from "classnames";

import Row, { IRow, IRowOptions } from "./row";
import { IIndexesMap, filterRowsByIndexes, getRowTreeLength, filterIndexes, IElevateds } from "../utils/table";
import { ISelection } from "../table-selection/selection-handler";
import { Nullable } from "../typing";

export enum Type {
  error = "error",
  success = "success",
  warning = "warning",
  info = "info"
}

export interface ITree {
  rowIndex: number;
  columnIndex: number;
  subTrees?: ITrees;
}

export interface ITrees {
  [rowIndex: number]: ITree;
}

export interface IColumnOptions {
  className?: string;
  size?: number;
  type?: Type;
  style?: React.CSSProperties;
}

export interface IColumn extends IColumnOptions {
  isSelectable?: boolean;
  disableLevelPadding?: boolean;
}

export interface IColumns {
  [index: number]: IColumn;
}

export interface IElementaryTable<IDataCoordinates = any> {
  id: string;
  /** All rows constituting the table */
  rows: IRow<IDataCoordinates>[];
  /** Specifies indexes of the columns to be shown */
  visibleColumnIndexes?: number[];
  /** Specifies indexes of the rows to be shown */
  visibleRowIndexes?: number[];
  /** Index of columns that need to be "elevated" by displaying a shadow on its right side */
  elevatedColumnIndexes?: IElevateds;
  /** Index of rows that need to be "elevated" by displaying a shadow on its bottom side */
  elevatedRowIndexes?: IElevateds;
  /** Options to customize any columns, such as size or align */
  columns?: IColumns;
  /** Options to customize any row, such as size */
  globalRowProps?: IRowOptions;
  /** Options to customize any column, such as size */
  globalColumnProps?: IColumnOptions;
  /** If true, we are using an additional column at the beginning of the rows to open/close the first openable child. */
  isSpan?: boolean;
}

export interface IElementaryTableProps<IDataCoordinates = any> extends IElementaryTable<IDataCoordinates>, ISelection {
  /** List of fixed rows on the top or bottom of your table. Delegated by the virtualizer component */
  fixedRowsIndexes?: number[];
  /** A mapping between relative and absolute rows indexes. Delegated by the Table component for example */
  indexesMapping: IIndexesMap;
  /** The list of the opend rows and sub-rows */
  openedTrees: ITrees;
  /** Called when a cell of the row is opened */
  onRowOpen?: (openedTree: ITree) => void;
  /** Called when a cell of the row is closed */
  onRowClose?: (closedTree: ITree) => void;
}

class ElementaryTable extends React.Component<IElementaryTableProps> {
  static defaultProps = {
    elevatedColumnIndexes: {},
    elevatedRowIndexes: {},
    openedTrees: {},
    selectedCells: {}
  };

  /** An utility of the table that return the length of the visible sub-rows
   * of the specified row by her absolute index. */
  public getRowTreeLength = (absoluteIndex: number): number => {
    const { visibleRowIndexes, indexesMapping } = this.props;
    // @ts-ignore we have a default value for visibleRowIndexes
    return getRowTreeLength(absoluteIndex, visibleRowIndexes || [], indexesMapping.absolute);
  };

  /** An utility of the table that return the visible rows for the specified row by her absolute index. */
  public getVisibleRows = (
    rows: IRow[],
    absoluteIndex: Nullable<number>,
    fixedRowsAbsoluteIndexes: number[] = []
  ): [number[], IRow[]] => {
    const { visibleRowIndexes, indexesMapping } = this.props;
    // @ts-ignore we have a default value for visibleRowIndexes
    return filterRowsByIndexes(rows, visibleRowIndexes || null, indexesMapping.absolute, absoluteIndex, fixedRowsAbsoluteIndexes);
  };

  private renderTableParts = (): {
    header: JSX.Element[];
    body: JSX.Element[];
  } => {
    const {
      id,
      isSpan,
      rows,
      columns,
      globalRowProps,
      globalColumnProps,
      visibleColumnIndexes,
      visibleRowIndexes,
      fixedRowsIndexes,
      indexesMapping,
      openedTrees,
      elevatedColumnIndexes,
      elevatedRowIndexes,
      onRowOpen,
      onRowClose,
      onCellMouseDown,
      onCellMouseUp,
      onCellMouseEnter,
      onCellContextMenu,
      selectedCells
    } = this.props;
    const [relativeIndexes, rowsToRender] = this.getVisibleRows(rows, null, fixedRowsIndexes);

    return rowsToRender.reduce<{ header: JSX.Element[]; body: JSX.Element[] }>(
      (result, row: IRow, index) => {
        if (!row) {
          return result;
        }
        const rowIndex = relativeIndexes ? relativeIndexes[index] : index;
        const { subItems, index: rowAbsoluteIndex } = indexesMapping.relative[rowIndex];
        const isVisible = !visibleRowIndexes || visibleRowIndexes.includes(rowAbsoluteIndex);
        // @ts-ignore we have a default value for openedTrees
        const rowOpenedTree = openedTrees[rowIndex];

        // @ts-ignore we have a default value for openedTrees
        const elevation = elevatedRowIndexes[rowAbsoluteIndex];
        let rowSelectedCells = (subItems || selectedCells[rowAbsoluteIndex]) && selectedCells;
        if (rowSelectedCells) {
          const nextRowMap = indexesMapping.relative && indexesMapping.relative[rowIndex + 1];
          const nextRowAbsoluteIndex = nextRowMap && nextRowMap.index;
          rowSelectedCells = nextRowAbsoluteIndex
            ? filterIndexes(rowSelectedCells, rowAbsoluteIndex, nextRowAbsoluteIndex)
            : rowSelectedCells;
        }
        const renderedRow = (
          <Row
            key={`row-${id}-${row.id}`}
            {...globalRowProps}
            {...row}
            className={classnames(row.className, {
              [`elevated-${elevation}`]: elevation
            })}
            absoluteIndex={rowAbsoluteIndex}
            index={rowIndex}
            isVisible={isVisible}
            isSpan={isSpan}
            columns={columns}
            elevatedColumnIndexes={elevatedColumnIndexes}
            globalColumnProps={globalColumnProps}
            visibleColumnIndexes={visibleColumnIndexes}
            visibleRowIndexes={visibleRowIndexes}
            openedTree={rowOpenedTree}
            relativeSubIndexesMapping={subItems}
            onOpen={onRowOpen}
            onClose={onRowClose}
            onCellMouseDown={onCellMouseDown}
            onCellMouseEnter={onCellMouseEnter}
            onCellMouseUp={onCellMouseUp}
            onCellContextMenu={onCellContextMenu}
            selectedCells={rowSelectedCells}
            // Table utils
            getVisibleRows={this.getVisibleRows}
            getRowTreeLength={this.getRowTreeLength}
          />
        );

        if (row.isHeader) {
          result.header.push(renderedRow);
        } else {
          result.body.push(renderedRow);
        }
        return result;
      },
      { header: [], body: [] }
    );
  };

  public render() {
    const { header, body } = this.renderTableParts();
    return (
      <table className="table-root">
        {header.length > 0 ? <thead>{header}</thead> : null}
        {body.length > 0 ? <tbody>{body}</tbody> : null}
      </table>
    );
  }
}

export default ElementaryTable;
