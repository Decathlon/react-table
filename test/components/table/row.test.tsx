/// <reference path="../../typings/tests-entry.d.ts" />
import * as React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { mount } from "enzyme";
import Row from "../../../src/components/table/row";
import { subRows, subMiam, tableWithDifferentRowSizes } from "../../../stories/utils/tables";
import { withThemeProvider } from "../../../stories/utils/decorators";
import { ITree } from "../../../src/components/table/elementary-table";

describe("Row component", () => {
  test("should render the default table", () => {
    const props = {
      id: "foo",
      cells: []
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Row {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render a simple row", () => {
    const props = subRows({})[0];
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Row {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render a Row with nested children", () => {
    const props = subRows({ subsubRows: subMiam })[0];
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Row {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render a Row with span", () => {
    const props = subRows({ subsubRows: subMiam })[0];
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Row {...props} isSpan rowSpanProps={{ title: "foo", color: "gray" }} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render a Row with a small size", () => {
    const props = tableWithDifferentRowSizes[2];
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Row {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should toggle the first cell", () => {
    const rows = subRows({ subsubRows: subMiam });
    const props = {
      index: 0,
      absolutIndex: 0,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      ...rows[0]
    };
    const RowComponent = ({ openedTree }: { openedTree?: ITree }) =>
      withThemeProvider(() => <Row {...props} openedTree={openedTree} />);
    const wrapper = mount(<RowComponent />);
    const instance: Row = wrapper.find(Row).instance() as Row;
    // @ts-ignore private method
    instance.toggleFirstCell();
    expect(props.onOpen).toBeCalledWith({ rowIndex: 0, columnIndex: 2 });
    wrapper.setProps({ openedTree: { rowIndex: 0, columnIndex: 2 } });
    // @ts-ignore private method
    instance.toggleFirstCell();
    expect(props.onClose).toBeCalledWith({ rowIndex: 0, columnIndex: 2 });
  });

  test("should toggle a cell", () => {
    const rows = subRows({ subsubRows: subMiam });
    const props = {
      index: 0,
      absolutIndex: 0,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      ...rows[0]
    };
    const RowComponent = ({ openedTree }: { openedTree?: ITree }) =>
      withThemeProvider(() => <Row {...props} openedTree={openedTree} />);
    const wrapper = mount(<RowComponent />);
    const instance: Row = wrapper.find(Row).instance() as Row;
    // @ts-ignore private method
    instance.toggleCell(2);
    expect(props.onOpen).toBeCalledWith({ rowIndex: 0, columnIndex: 2 });
    wrapper.setProps({ openedTree: { rowIndex: 0, columnIndex: 2 } });
    // @ts-ignore private method
    instance.toggleCell(2);
    expect(props.onClose).toBeCalledWith({ rowIndex: 0, columnIndex: 2 });
  });

  test("should open subRow", () => {
    const rows = subRows({ subsubRows: subMiam });
    const row = rows[0];
    // @ts-ignore
    row.cells[2].subItems[0].cells[2].subItems = subMiam;
    const props = {
      index: 0,
      absolutIndex: 0,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      ...row
    };
    const RowComponent = ({ openedTree }: { openedTree?: ITree }) =>
      withThemeProvider(() => <Row {...props} openedTree={openedTree} />);
    const wrapper = mount(<RowComponent />);
    const instance: Row = wrapper.find(Row).instance() as Row;
    wrapper.setProps({ openedTree: { rowIndex: 0, columnIndex: 2 } });
    // @ts-ignore private method
    instance.onSubRowOpen({ rowIndex: 0, columnIndex: 2 });
    expect(props.onOpen).toBeCalledWith({
      rowIndex: 0,
      columnIndex: 2,
      subTrees: { 0: { rowIndex: 0, columnIndex: 2 } }
    });
  });

  test("should close subRow", () => {
    const rows = subRows({ subsubRows: subMiam });
    const row = rows[0];
    // @ts-ignore
    row.cells[2].subItems[0].cells[2].subItems = subMiam;
    const props = {
      index: 0,
      absolutIndex: 0,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      ...row
    };
    const RowComponent = ({ openedTree }: { openedTree?: ITree }) =>
      withThemeProvider(() => <Row {...props} openedTree={openedTree} />);
    const wrapper = mount(<RowComponent />);
    const instance: Row = wrapper.find(Row).instance() as Row;
    wrapper.setProps({
      openedTree: {
        rowIndex: 0,
        columnIndex: 2,
        subTrees: { 0: { rowIndex: 0, columnIndex: 2 } }
      }
    });
    // @ts-ignore private method
    instance.onSubRowClose({ rowIndex: 0, columnIndex: 2 });
    expect(props.onOpen).toBeCalledWith({
      rowIndex: 0,
      columnIndex: 2,
      subTrees: {}
    });
  });
});
