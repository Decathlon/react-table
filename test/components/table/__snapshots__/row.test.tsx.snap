// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row component should render a Row with a small size 1`] = `
<React.Fragment>
  <tr
    className="table-row"
    data-testid="table-row-small"
    style={
      Object {
        "height": 25,
      }
    }
  >
    <Cell
      className=""
      colspan={1}
      component="td"
      id="small"
      index={0}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 25,
          "right": 0,
          "width": "auto",
        }
      }
      value="Wawooo!"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="wah"
      index={1}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 25,
          "right": 0,
          "width": "auto",
        }
      }
      value="Wah!"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="cool"
      index={2}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 25,
          "right": 0,
          "width": "auto",
        }
      }
      value="Cool!"
    />
  </tr>
</React.Fragment>
`;

exports[`Row component should render a Row with nested children 1`] = `
<React.Fragment>
  <tr
    className="table-row"
    data-testid="table-row-foo"
    style={
      Object {
        "height": 56,
      }
    }
  >
    <Cell
      className=""
      colspan={1}
      component="td"
      id="storybook"
      index={0}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Storybook"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="jest"
      index={1}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Jest"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="Enzym"
      index={2}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      subItems={
        Array [
          Object {
            "cells": Array [
              Object {
                "id": "sushi",
                "value": "Sushi",
              },
              Object {
                "id": "pizza",
                "value": "Pizza",
              },
              Object {
                "id": "mafe",
                "value": "Mafe",
              },
            ],
            "id": "miam",
          },
        ]
      }
      value="enzyme"
    />
  </tr>
</React.Fragment>
`;

exports[`Row component should render a Row with span 1`] = `
<React.Fragment>
  <tr
    className="table-row"
    data-testid="table-row-foo"
    style={
      Object {
        "height": 56,
      }
    }
  >
    <RowSpan
      color="gray"
      length={1}
      opened={false}
      title="foo"
      toggle={[Function]}
      width={110}
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      hideSubItemsOpener={false}
      id="storybook"
      index={0}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Storybook"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      hideSubItemsOpener={false}
      id="jest"
      index={1}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Jest"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      hideSubItemsOpener={true}
      id="Enzym"
      index={2}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      subItems={
        Array [
          Object {
            "cells": Array [
              Object {
                "id": "sushi",
                "value": "Sushi",
              },
              Object {
                "id": "pizza",
                "value": "Pizza",
              },
              Object {
                "id": "mafe",
                "value": "Mafe",
              },
            ],
            "id": "miam",
          },
        ]
      }
      value="enzyme"
    />
  </tr>
</React.Fragment>
`;

exports[`Row component should render a simple row 1`] = `
<React.Fragment>
  <tr
    className="table-row"
    data-testid="table-row-foo"
    style={
      Object {
        "height": 56,
      }
    }
  >
    <Cell
      className=""
      colspan={1}
      component="td"
      id="storybook"
      index={0}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Storybook"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="jest"
      index={1}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      value="Jest"
    />
    <Cell
      className=""
      colspan={1}
      component="td"
      id="Enzym"
      index={2}
      isSelectable={true}
      isSelected={false}
      onCallOpen={[Function]}
      opened={false}
      relativeRowIndex={0}
      rowIndex={0}
      style={
        Object {
          "height": 56,
          "right": 0,
          "width": "auto",
        }
      }
      subItems={Array []}
      value="enzyme"
    />
  </tr>
</React.Fragment>
`;

exports[`Row component should render the default table 1`] = `
<React.Fragment>
  <tr
    className="table-row"
    data-testid="table-row-foo"
    style={
      Object {
        "height": 56,
      }
    }
  />
</React.Fragment>
`;
